core/src/time.rs

```diff
@@ -221,14 +232,117 @@ impl TimeExt for Time {
 }
 
 /// Extension methods on [`Interval`].
-pub trait IntervalExt {
+pub trait IntervalExt: Sized {
     /// Returns a [`Time`] representing the excluded end of this interval.
     fn end(&self) -> Time;
+
+    /// Returns a new minimal [`Interval`] that contains both this interval and `other`.
+    fn merge(&self, other: &Self) -> Result<Self, Error>;
+
+    /// Returns a 0-length `[Interval]` that contains exactly the provided [`Time`].
+    fn from_time(time: &Time) -> Result<Self, Error>;
 }
 
 impl IntervalExt for Interval {
     fn end(&self) -> Time {
         // [`Self::new`] verified that this addition doesn't overflow.
         self.start().add(self.duration()).unwrap()
     }
+
+    fn merge(&self, other: &Self) -> Result<Self, Error> {
+        let max_time = std::cmp::max(self.end(), other.end());
+        let min_time = std::cmp::min(self.start(), other.start());
+
+        // This can't actually fail for any valid Intervals
+        Self::new(*min_time, max_time.difference(min_time)?)
+    }
+
+    fn from_time(time: &Time) -> Result<Self, Error> {
```

Comment:
Arguably this should be `impl From<Time> for Interval`, but that can't be defined in `janus_core` due to the orphan rule.